File: src/utils/diagnostic.cpp
#include "utils/diagnostic.hpp"
#include <iostream>

using namespace utils;

Diagnostic::Diagnostic(DiagnosticKind kind, const std::string &message)
    : kind(kind), message(message), position(std::nullopt) {

}

Diagnostic::Diagnostic(DiagnosticKind kind, const std::string &message, const Position &position)
    : kind(kind), message(message), position(position) {

}

Diagnostic::Diagnostic(common::values::ValueError *error, const Position &position)
    : kind(DiagnosticKind::Error), position(position) {
    this->message = error->getName().value_or("");
    if (error->getDetails())
        this->message += ": " + error->getDetails().value();
}

void Diagnostic::show(const SourceText &source) const {
    std::string color = Diagnostic::getColorByKind(this->kind);
    if (this->position)
        this->showWithPosition(color, source);
    else
        this->showWithoutPosition(color);
}

DiagnosticKind Diagnostic::getKind() const {
    return this->kind;
}

void Diagnostic::showWithoutPosition(const std::string &color) const {
    std::cout << color << this->message << ".\x1b[0m\n";
}

void Diagnostic::showWithPosition(const std::string &color, const SourceText &source) const {
    SourceInformations infos;
    source.getInfosForPosition(this->position.value(), infos);
    if (infos.line_start == infos.line_end) {
        std::optional<std::string> line_text = source.getLineText(infos.line_start);
        if (!line_text)
            return;
        
        std::cout << line_text.value() << '\n' 
            << color << std::string(infos.left_offset, ' ') << std::string(this->position.value().getUtf8Length(), '^') << '\n'
            << "[At line " << infos.line_start + 1 << ", column " << infos.left_offset + 1 << '-' << infos.left_offset + this->position.value().getUtf8Length() << "]: " << this->message << ".\x1b[0m\n";
    } else {
        for (uint64_t i = infos.line_start; i <= infos.line_end; i++) {
            std::optional<std::string> line_text = source.getLineText(i);
            std::optional<uint64_t> line_length = source.getLineLength(i);
            if (!line_text || !line_length)
                return;
            
            std::cout << line_text.value() << '\n';
            if (i == infos.line_start) {
                std::cout << color << std::string(infos.left_offset, ' ') << std::string(line_length.value() - infos.left_offset, '^') << "\x1b[0m\n";
            } else if (i == infos.line_end) {
                std::cout << color << std::string(line_length.value() - infos.right_offset, '^') << "\x1b[0m\n";
            } else {
                std::cout << color << std::string(line_length.value(), '^') << "\x1b[0m\n";
            }
        }

        std::cout << color << "[From line " << infos.line_start + 1 << " to line " << infos.line_end + 1 << "]: " << this->message << ".\x1b[0m\n";
    }
}

std::string Diagnostic::getColorByKind(DiagnosticKind kind) {
    switch (kind) {
        case DiagnosticKind::Success: return "\x1b[32m";
        case DiagnosticKind::Warning: return "\x1b[33m";
        case DiagnosticKind::Error: return "\x1b[31m";

        case DiagnosticKind::Information: default: return "\x1b[37m";
    }
}

File: src/utils/diagnostic.rs
use crate::utils::position::Position;
use crate::utils::source::SourceText;

#[derive(PartialEq)]
pub enum DiagnosticKind {
    Success,
    Info,
    Warning,
    Error,
}

pub struct Diagnostic {
    kind: DiagnosticKind,
    message: String,
    position: Option<Position>,
}

impl Diagnostic {
    pub fn init_message(kind: DiagnosticKind, message: String) -> Diagnostic {
        Diagnostic { kind, message, position: None }
    }

    pub fn init_complete(kind: DiagnosticKind, message: String, position: Position) -> Diagnostic {
        Diagnostic { kind, message, position: Some(position) }
    }

    pub fn get_kind(&self) -> &DiagnosticKind {
        &self.kind
    }

    pub fn show(&self, source: &SourceText) {
        let color = Self::get_color_by_kind(&self.kind);
        if let Some(position) = &self.position {
            self.show_with_position(color, source, position);
        } else {
            self.show_without_position(color);
        }
    }

    fn show_without_position(&self, color: &str) {
        println!("{color}{}.\x1b[0m", self.message);
    }

    fn show_with_position(&self, color: &str, source: &SourceText, position: &Position) {
        let mut line_start: usize = 0;
        let mut line_end: usize = 0;
        let mut left_offset: usize = 0;
        let mut right_offset: usize = 0;
        source.get_infos_for_position(position, &mut line_start, &mut line_end, &mut left_offset, &mut right_offset);

        if line_start == line_end {
            let line_text = source.get_line_text(line_start).unwrap();
            let offset_text = " ".repeat(left_offset);
            let arrow_text = "^".repeat(position.get_length());

            println!("{line_text}\n{color}{offset_text}{arrow_text}\n[At line {}, column {}-{}]: {}.\x1b[0m", line_start + 1, left_offset + 1, left_offset + position.get_length(), self.message);
        } else {
            for i in line_start..line_end+1 {
                let line_text = source.get_line_text(i).unwrap();
                let line_length = source.get_line_length(i).unwrap();
                println!("{line_text}");

                if i == line_start {
                    let offset_text = " ".repeat(left_offset);
                    let arrow_text = "^".repeat(line_length - left_offset);
                    println!("{color}{offset_text}{arrow_text}\x1b[0m");
                } else if i == line_end {
                    let arrow_text = "^".repeat(line_length - right_offset);
                    println!("{color}{arrow_text}\x1b[0m");
                } else {
                    let arrow_text = "^".repeat(line_length);
                    println!("{color}{arrow_text}\x1b[0m");
                }
            }

            println!("{color}[From line {} to line {}]: {}.\x1b[0m", line_start + 1, line_end + 1, self.message);
        }
    }

    fn get_color_by_kind(kind: &DiagnosticKind) -> &str {
        match kind {
            DiagnosticKind::Success => "\x1b[32m",
            DiagnosticKind::Info => "\x1b[37m",
            DiagnosticKind::Warning => "\x1b[33m",
            DiagnosticKind::Error => "\x1b[31m",
        }
    }
}